### PR TITLE
Add types to the rvsdg and egglog encoding

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

File: src/rvsdg/mod.rs [hunk 1]
```diff
@@ -92,17 +92,17 @@ pub(crate) type Id = usize;
 #[derive(Debug, PartialEq)]
 pub(crate) enum Expr<Op> {
```
Comment (ezrosent) [inline src/rvsdg/mod.rs:93]: optional: should we name the fields now that there are so many (similar to the various `Ops` in the bril Rust bindings?)

File: src/rvsdg/from_cfg.rs [hunk 2]
```diff
@@ -75,12 +79,15 @@ pub(crate) fn cfg_func_to_rvsdg(cfg: &mut Cfg) -> Result<RvsdgFunction> {
     })
 }
 
+pub(crate) type FunctionTypes = HashMap<String, Option<Type>>;
```
Comment (ezrosent) [inline src/rvsdg/from_cfg.rs:82]: Why Option<Type> (and document it here)

File: src/rvsdg/mod.rs [hunk 3]
```diff
@@ -209,28 +214,38 @@ pub enum EgglogFunctionResult {
...
-        };
-        fn from_ty(ty: &Type) -> egglog::ast::Expr {
-            match ty {
-                Type::Int => Call("IntT".into(), vec![]),
-                Type::Bool => Call("BoolT".into(), vec![]),
-                Type::Float => Call("FloatT".into(), vec![]),
-                Type::Char => Call("CharT".into(), vec![]),
-                Type::Pointer(ty) => Call("PointerT".into(), vec![from_ty(ty.as_ref())]),
+        let f = |operands: &Vec<Operand>, ty: Option<Type>| {
+            let mut res = vec![];
```
Comment (ezrosent) [inline src/rvsdg/mod.rs:231]: nit: `Vec::with_capacity(operands.len() + ty.is_some() as usize)` (or similar)

**Review by ezrosent — APPROVED** [review]
Let me know if you want another round of feedback; but I'll leave it up to you if you want to merge after responding.

File: src/rvsdg/mod.rs (hunk 1 above)
Comment (oflatt) [reply src/rvsdg/mod.rs:93]: We probably should (I wish there was a cool vscode refactor for this)

File: src/rvsdg/from_cfg.rs (hunk 2 above)
Comment (oflatt) [reply src/rvsdg/from_cfg.rs:82]: Left a comment- functions don't always return a value so for these we use `None`. Bril doesn't have a void type

**Pull request merged**